### PR TITLE
Handle predicates

### PR DESCRIPTION
**Pull request by PKuebler**
Hello,

I have made a first attempt to teach the module predicates. I used the javascript example from issue #77 as a guideline.

Unfortunately the structure differs between the `go-tree-sitter` and the NodeJS version. In the NodeJS version the text is already present at each node.

To avoid breaking changes, I put it into a separate function that has to be executed separately.

The code changes are divided into 2 areas. The validation of the query takes place during the query creation, because there is an error handling. The application takes place later.

**Still needed input**
- What is the best way to do the error handling. The existing error handling does not provide the space for information.
- I am having a hard time adding tests since the previous tests generate anything. Would it be ok if I define test cases with their own input?

I am very uncomfortable that I have to ask you this, would it be possible to add the label `hacktoberfest-accepted` to the pull request?

**Example**
```golang
match, _ := qc.NextMatch()
match = qc.FilterPredicates(match, src)
if len(m.Captures) == 0 {
	continue
}
``` 

**Complete Example**
```golang
parser := sitter.NewParser()
parser.SetLanguage(javascript.GetLanguage())

fmt.Printf("Start with %s\n", test.Src)

src, err := os.ReadFile("test.js")
if err != nil {
	panic(err)
}

query, err := os.ReadFile("query.scm")
if err != nil {
	panic(err)
}

tree := parser.Parse(nil, src)
n := tree.RootNode()

q, _ := sitter.NewQuery(query, javascript.GetLanguage())
qc := sitter.NewQueryCursor()
qc.Exec(q, n)

for {
	m, ok := qc.NextMatch()
	if !ok {
		break
	}

	m = qc.FilterPredicates(m, src)
	if len(m.Captures) == 0 {
		continue
	}

	fmt.Printf(result, "  pattern: %d\n", m.PatternIndex)

	for _, c := range m.Captures {
		captureName := q.CaptureNameForId(c.Index)

		fmt.Printf(result, "    capture: %d", c.Index)
		fmt.Printf(result, " - %s", captureName)
		start := c.Node.StartPoint()
		end := c.Node.EndPoint()
		text := c.Node.Content(src)
		fmt.Printf(result, ", start: (%d, %d), end: (%d, %d), text: `%s`\n", start.Row, start.Column, end.Row, end.Column, text)
	}
}
```

### CODE REVIEW

**Comment by smacker** [comment]
hi @PKuebler! great work! let me review it in detail later this week.

File: bindings.go [hunk 1]
```diff
@@ -737,6 +738,49 @@ func NewQuery(pattern []byte, lang *Language) (*Query, error) {
...
 	q := &Query{c: c}
+
+	for i := uint32(0); i < q.PatternCount(); i++ {
+		steps := q.PredicatesForPattern(i)
+		if len(steps) == 0 {
+			continue
+		}
+
+		if steps[0].Type != QueryPredicateStepTypeString {
+			panic("preidcates must begin with a literal value")
```
Comment (smacker) [inline bindings.go:749]: `preidcates` -> `predicate`

Comment (smacker) [reply bindings.go:749]: let's avoid `panic` and return an error here instead

File: bindings.go [hunk 2]
```diff
@@ -737,6 +738,49 @@ func NewQuery(pattern []byte, lang *Language) (*Query, error) {
...
+		}
+
+		operator := q.StringValueForId(steps[0].ValueId)
+		switch operator {
+		case "eq?", "not-eq?":
+			if len(steps) != 4 {
+				panic(fmt.Sprintf("Wrong number of arguments to `#%s` predicate. Expected 2, got %d", operator, len(steps)-1))
```
Comment (smacker) [inline bindings.go:756]: here and in all the cases below let's replace panic with returning an error

File: bindings.go [hunk 3]
```diff
@@ -737,6 +738,49 @@ func NewQuery(pattern []byte, lang *Language) (*Query, error) {
...
 
 	q := &Query{c: c}
+
+	for i := uint32(0); i < q.PatternCount(); i++ {
+		steps := q.PredicatesForPattern(i)
+		if len(steps) == 0 {
+			continue
+		}
+
+		if steps[0].Type != QueryPredicateStepTypeString {
```
Comment (smacker) [inline bindings.go:748]: do we need to iterate over steps instead of getting the first one only?

File: bindings.go [hunk 4]
```diff
@@ -939,6 +983,83 @@ func (qc *QueryCursor) NextCapture() (*QueryMatch, uint32, bool) {
...
+		PatternIndex: m.PatternIndex,
+	}
+
+	steps := qc.q.PredicatesForPattern(uint32(qm.PatternIndex))
+	if len(steps) == 0 {
+		qm.Captures = m.Captures
+		return qm
+	}
+
+	operator := qc.q.StringValueForId(steps[0].ValueId)
```
Comment (smacker) [inline bindings.go:998]: do we need to iterate over steps instead of getting the first one only?

**Comment by smacker** [comment]
> What is the best way to do the error handling. The existing error handling does not provide the space for information.

I see you `panic` in the code. The signature of the functions allows passing an error in the second argument. What is your concern about using it?

> I am having a hard time adding tests since the previous tests generate anything. Would it be ok if I define test cases with their own input?

Yes, sure. You can define your own test cases.

Thanks again for taking this initiative!

File: bindings.go (hunk 1 above)
Comment (PKuebler) [reply bindings.go]: see below

**Comment by PKuebler** [comment]
I got more into test grammar and was able to find ways to write the tests with it after all.

**Comment by smacker** [comment]
This is amazing work. Thank you @PKuebler so much!

**Pull request merged**